docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,454 @@
+will not be read back by future generations of pageservers.
+
+The _AttachedStale_ state simply disables S3 uploads. The stale pageserver
+will continue to ingest the WAL and write layers to local disk, but not to
+do any uploads to S3.
+
+A node may enter AttachedStale in two ways:
+
+- Explicitly, when control plane calls into the node at the start of a migration.
+- Implicitly, when the node tries to validate some deletions and discovers
+  that its generation is tale.
+
+Over long periods of time, a node in AttachedStale will accumulate data
+on local disk, as it cannot evict any layers written since it entered the
+AttachStale state. We rely on the control plane to revert the location to
+Secondary or Detached at the end of a migration.
+
+### Warm secondary updates
+
+#### Layer heatmap
+
+The secondary location's job is to serve reads **with the same quality of service as the original location
+was serving them around the time of a migration**. This does not mean the secondary
+location needs the whole set of layers: inactive layers that might soon
+be evicted on the attached pageserver need not be downloaded by the
+secondary. A totally idle tenant only needs to maintain enough on-disk
+state to enable a fast cold start (i.e. the most recent image layers are
+typically sufficient).
+
+To enable this, we introduce the concept of a _layer heatmap_, which
+acts as an advisory input to secondary locations to decide which
+layers to download from S3.
+
+#### Attached pageserver
+
+The attached pageserver, if in state AttachedSingle, periodically
+uploads a serialized heat map to S3. It may skip this if there
+is no change since the last time it uploaded (e.g. if the tenant
+is totally idle).
+
+#### Secondary pageserver behavior
+
+Secondary warm locations run a simple loop, implemented separately from
+the main `Tenant` type, which represents attached tenants:
+
+- Download the layer heatmap
+- Select any "hot enough" layers to download, if there is sufficient
+  free disk space.
+- Download layers
+- Download the latest index_part.json
+- Check if any layers currently on disk are no longer referenced by
+  the tenant's metadata & delete them
+
+### Location configuration API
+
+The pageserver will expose a new per-tenant API for setting
+the state: `/tenant/<tenant_id>/configure`.
+
+Body content:
+
+```
+{
+  state: 'enum{Detached, Secondary, AttachedSingle, AttachedMulti, AttachedStale}',
+  generation: Option<u32>,
+  configuration: `Option<TenantConfig>`
+  flush: bool
+}
+```
+
+Existing `/attach` and `/detach` endpoint will have the same
+behavior as calling `/configure` with `AttachedSingle` and `Detached`
+states respectively.
+
+The generation attribute is mandatory for entering `AttachedSingle` or
+`AttachedMulti`.
+
+The configuration attribute is mandatory when entering any state other
+than `Detached`.
```

Comment:
Let's model this as kubernetes-style algebraic types where all fields are mandatory?

```json

state:
  AttachedMulti:
    generation: 
    configuration: 
  ...
```

Reply:
I think it becomes less readable that way: one has to read the whole spec to grok how each field is used, rather than just having it in prose -- the RFC is meant as an explanatory doc for humans.

Reply:
>I think it becomes less readable that way

I don't. The reason is that you have several "is mandatory only in" in this section, but, you never explain what happens if the field isn't specified, leaving the reader to wonder/worry about whether there's a missed corner case.

By sticking to the kubernetes-style algebraic types, you reduce the value space that the reader has to think  about.

Reply:
Given that you and I already understand it, at the point we are asking third parties to grok the configuration states I think we can point them at the Rust definition of `LocationConf` as the master definition of tenant states.  I don't think it's worthwhile to spend time crafting an in-rfc version of that, since you and I are the ones reviewing this doc & we both already understand it.